src/crimson/os/seastore/lba_manager/btree/btree_range_pin.cc

```diff
@@ -0,0 +1,134 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/common/log.h"
+
+#include "crimson/os/seastore/lba_manager/btree/btree_range_pin.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore::lba_manager::btree {
+
+btree_range_pin_t::~btree_range_pin_t()
+{
+  assert(!pins == !is_linked());
+  assert(!ref);
+  if (pins) {
+    logger().debug("{}: removing {}", __func__, *this);
+    pins->remove_pin(*this, true);
+  }
+  extent = nullptr;
+}
+
+void btree_pin_set_t::remove_pin(btree_range_pin_t &pin, bool do_check_parent)
+{
+  logger().debug("{}: {}", __func__, pin);
+  assert(pin.is_linked());
+  assert(pin.pins);
+  assert(!pin.ref);
+
+  pins.erase(pin);
+  pin.pins = nullptr;
+
+  if (do_check_parent) {
+    check_parent(pin);
+  }
+}
+
+btree_range_pin_t *btree_pin_set_t::maybe_get_parent(
+  const lba_node_meta_t &meta)
+{
+  auto cmeta = meta;
+  cmeta.depth++;
+  auto iter = pins.upper_bound(cmeta, btree_range_pin_t::meta_cmp_t());
+  if (iter == pins.begin()) {
+    return nullptr;
+  } else {
+    --iter;
+    if (iter->range.is_parent_of(meta)) {
+      return &*iter;
+    } else {
+      return nullptr;
+    }
+  }
+}
+
+btree_range_pin_t *btree_pin_set_t::maybe_get_first_child(
+  const lba_node_meta_t &meta)
+{
+  if (meta.depth == 0) {
+    return nullptr;
+  }
+
+  auto cmeta = meta;
+  cmeta.depth--;
+
+  auto iter = pins.lower_bound(cmeta, btree_range_pin_t::meta_cmp_t());
+  if (iter == pins.end()) {
+    return nullptr;
+  } else if (meta.is_parent_of(iter->range)) {
+    return &*iter;
+  } else {
+    return nullptr;
+  }
+}
+
+void btree_pin_set_t::release_if_no_children(btree_range_pin_t &pin)
+{
+  assert(pin.is_linked());
+  if (maybe_get_first_child(pin.range) == nullptr) {
+    pin.drop_ref();
+  }
+}
+
+void btree_pin_set_t::add_pin(btree_range_pin_t &pin)
+{
+  assert(!pin.is_linked());
+  assert(!pin.pins);
+  assert(!pin.ref);
+
+  auto [prev, inserted] = pins.insert(pin);
+  if (!inserted) {
+    logger().debug("{}: unable to add {}, found {}", __func__, pin, *prev);
+    assert(0 == "impossible");
+    return;
+  }
+  pin.pins = this;
+  if (!pin.is_root()) {
+    auto *parent = maybe_get_parent(pin.range);
+    assert(parent);
+    if (!parent->has_ref()) {
```

Comment:
i don't quite understand this. there are two cases: LBANode (depth >= 1) and LogicalCachedExtent (depth == 0). and keep all the ancestors of a node are kept in the pin set. also, i assume that we always load an LBANode's parent before loading it. and all nodes tracked by `btree_range_pin_t` should have a valid `ref` . unless we are in the middle of `BtreeLBAManager::complete_transaction()`, where the pin_set is updated and the invariant could be violated before that function returns.

but since we add_pin in the order from parent to leaf in `BtreeLBAManager::complete_transaction()`, the parent should always has a valid ref, i think.

Reply:
No, until a node has a child, this structure does not maintain a refcount to it (it maintains a *reference*, but not a refcount).  In general, when we add a node, we need to acquire a ref on the parent unless it already has one.

Reply:
That is, we add nodes from root to leaf LogicalCachedExtent.  Each added node necessarily does not have children (except during complete_transaction()), and therefore is not pinned by this structure (that is, the structure contains a reference to that node, but does not hold a refcount).  It's only once we then add the child node that we actually pin the parent.

Reply:
i see.